### ISSUE TITLE
Questions on concurrency scenarios

### ISSUE THREAD

**Issue by JohannesRudolph** [meshcloud/unipipe-service-broker#24]
So @felixzieger and me worked a ton with unipipe yesterday and I think we discovered an issue - just want to quickly check in with the experts @j0g3sc and @sttomm  if we’re on the right track or misunderstood something about unipipe’s design. 

So we encountered an 

```
-8075-exec-3] .s.c.ServiceBrokerWebMvcExceptionHandler : Unknown exception handled: 
org.eclipse.jgit.api.errors.WrongRepositoryStateException: Cannot pull into a repository with state: MERGING
	at org.eclipse.jgit.api.PullCommand.call(PullCommand.java:244) ~[org.eclipse.jgit-4.11.0.201803080745-r.jar!/:4.11.0.201803080745-r]
	at io.meshcloud.dockerosb.persistence.SynchronizedGitHandler.internalPull(SynchronizedGitHandler.kt:238) ~[classes!/:na]
	at 
```

So I had a quick take on the code and that resulted in a couple of questions: 

1. We were modifying the repo from the CI pipeline and OSB concurrently - this should be supported, right? 
2. I see that unipipe internalPull uses git pull - but git pull will always merge (unless you do git pull --rebase  or git pull --ff-only - so I assume that’s why the repo went into MERGING state in the first place? I think we also need to apply our rebase loop sort of logic when pulling…. 
3. Speaking of the rebase loop I see another issue - none of the controller handlers that writes yaml files to the repo synchronizes its writes to the repo with other git operations - this means that a controller can attempt to write an `instance.yml` while the `ScheduledPushHandler` is in the middle of running a rebase loop. This can lead to undesirable side-effects.

**Comment by sttomm** [comment]
> So @felixzieger and me worked a ton with unipipe yesterday and I think we discovered an issue - just want to quickly check in with the experts @j0g3sc and @sttomm if we’re on the right track or misunderstood something about unipipe’s design.
> 
> So we encountered an
> 
> ```
> -8075-exec-3] .s.c.ServiceBrokerWebMvcExceptionHandler : Unknown exception handled: 
> org.eclipse.jgit.api.errors.WrongRepositoryStateException: Cannot pull into a repository with state: MERGING
> 	at org.eclipse.jgit.api.PullCommand.call(PullCommand.java:244) ~[org.eclipse.jgit-4.11.0.201803080745-r.jar!/:4.11.0.201803080745-r]
> 	at io.meshcloud.dockerosb.persistence.SynchronizedGitHandler.internalPull(SynchronizedGitHandler.kt:238) ~[classes!/:na]
> 	at 
> ```
> 
> So I had a quick take on the code and that resulted in a couple of questions:
> 
> 1. We were modifying the repo from the CI pipeline and OSB concurrently - this should be supported, right?
> 2. I see that unipipe internalPull uses git pull - but git pull will always merge (unless you do git pull --rebase  or git pull --ff-only - so I assume that’s why the repo went into MERGING state in the first place? I think we also need to apply our rebase loop sort of logic when pulling….
> 3. Speaking of the rebase loop I see another issue - none of the controller handlers that writes yaml files to the repo synchronizes its writes to the repo with other git operations - this means that a controller can attempt to write an `instance.yml` while the `ScheduledPushHandler` is in the middle of running a rebase loop. This can lead to undesirable side-effects.

1. yes, this should be supported
2. sounds reasonable
3. sounds reasonable and your solution in the linked PR sounds like a good solution to it

**Comment by JohannesRudolph** [comment]
Closed via #25
